packages/storage-plus/src/snapshot/map.rs

```diff
@@ -67,52 +85,64 @@ where
 
 impl<'a, K, T> SnapshotMap<'a, K, T>
 where
-    T: Serialize + DeserializeOwned + Clone,
-    K: PrimaryKey<'a> + Prefixer<'a> + KeyDeserialize,
+    K: Serialize + DeserializeOwned + PartialOrd + Clone,
+    T: Serialize + DeserializeOwned,
 {
-    pub fn key(&self, k: K) -> Path<T> {
-        self.primary.key(k)
-    }
-
-    fn no_prefix_raw(&self) -> Prefix<Vec<u8>, T, K> {
-        self.primary.no_prefix_raw()
-    }
-
     /// load old value and store changelog
     fn write_change(&self, store: &mut dyn Storage, k: K, height: u64) -> StdResult<()> {
         // if there is already data in the changelog for this key and block, do not write more
-        if self.snapshots.has_changelog(store, k.clone(), height)? {
+        // @todo we need to serialize the key like in KeyMap
+        if self
+            .snapshots
+            .has_changelog(store, &self.serialize_key(&k)?, height)?
+        {
             return Ok(());
         }
         // otherwise, store the previous value
-        let old = self.primary.may_load(store, k.clone())?;
-        self.snapshots.write_changelog(store, k, height, old)
+        let old = self.may_load(store, k.clone())?;
+        self.snapshots
+            .write_changelog(store, &self.serialize_key(&k)?, height, old)
     }
 
     pub fn save(&self, store: &mut dyn Storage, k: K, data: &T, height: u64) -> StdResult<()> {
-        if self.snapshots.should_checkpoint(store, &k)? {
+        if self
+            .snapshots
+            .should_checkpoint(store, &self.serialize_key(&k)?)?
+        {
             self.write_change(store, k.clone(), height)?;
         }
-        self.primary.save(store, k, data)
+        self.primary.insert(store, &k, &data)?;
+        match self.key_index.insert(store, &k) {
+            Err(StdError::GenericErr { .. }) => Ok(()), // just means element already exists
+            Err(e) => Err(e),                           // real error
+            _ => Ok(()),
+        }
     }
 
     pub fn remove(&self, store: &mut dyn Storage, k: K, height: u64) -> StdResult<()> {
-        if self.snapshots.should_checkpoint(store, &k)? {
+        if self
+            .snapshots
+            .should_checkpoint(store, &self.serialize_key(&k)?)?
+        {
             self.write_change(store, k.clone(), height)?;
         }
-        self.primary.remove(store, k);
-        Ok(())
+        self.primary.remove(store, &k)
+        // Here we would like to remove the corresponding entry in our key index BST, but that
+        // operation doesn't exist so we will just leave it. In extreme cases it will slow down
+        // iteration, but in practice it's probably negligible.
     }
 
     /// load will return an error if no data is set at the given key, or on parse error
     pub fn load(&self, store: &dyn Storage, k: K) -> StdResult<T> {
-        self.primary.load(store, k)
+        self.primary
+            .get(store, &k)
+            .ok_or(StdError::not_found("key"))
```

Comment:
use `.ok_or_else()` when unwrapping to a function call
```suggestion
            .ok_or_else(|| StdError::not_found("key"))
```